common/ProxyLookupCache.go

```diff
@@ -0,0 +1,166 @@
+	lookupTimeout   time.Duration
+	lookupLock      *sync.Mutex
+	lookupMethod    ProxyLookupFunc
+}
+
+func (c *proxyLookupCache) getProxyNoCache(req *http.Request) proxyLookupResult {
+	// Look up in background, so we can set timeout
+	// We do this since we have observed cases that don't return on Windows 10, presumably due to OS issue as described here:
+	// https://developercommunity.visualstudio.com/content/problem/282756/intermittent-and-indefinite-wcf-hang-blocking-requ.html
+	ch := make(chan proxyLookupResult)
+	go func() {
+		u, err := c.lookupMethod(req) // typically configured to call the result of ieproxy.GetProxyFunc
+		ch <- proxyLookupResult{u, err}
+	}()
+
+	select {
+	case v := <-ch:
+		return v
+	case <-time.After(c.lookupTimeout):
+		return proxyLookupResult{nil, ProxyLookupTimeoutError}
+		// Note: in testing the the real app, this code path wasn't triggered. Not sure if its just luck that in many Win10 test runs,
+		// with hundreds of thousands of files each, this didn't trigger - even though the underlying issue did trigger
+		// on about 25% of similar test runs prior to this code being added. Maybe just luck, or maybe something about spinning up
+		// the separate goroutine here as "magically" prevented the issue.
+	}
+}
+
+// getProxy returns the cached proxy, or looks it up if its not already cached
+func (c *proxyLookupCache) getProxy(req *http.Request) (*url.URL, error) {
+	var value proxyLookupResult
+	var ok bool
+
+	// key is the scheme+host portion of the URL
+	key := url.URL{
+		Scheme: req.URL.Scheme,
+		User:   req.URL.User,
+		Host:   req.URL.Host,
+	}
+
+	// if we've got it return it
+	if value, ok = c.mapLoad(key); ok {
+		return value.url, value.err
+	}
+
+	// else, look it up with the (potentially expensive) lookup function
+	// Because the function is potentially expensive(ish) and because we only want to kick off one refresh GR per key,
+	// we use a lock here
+	c.lookupLock.Lock()
+	defer c.lookupLock.Unlock()
+
+	if value, ok = c.mapLoad(key); !ok {
+		value = c.getProxyNoCache(req)
+		c.m.Store(key, value)
+		go c.endlessTimedRefresh(key, req)
+	}
+
+	return value.url, value.err
+}
+
+func (c *proxyLookupCache) mapLoad(key url.URL) (proxyLookupResult, bool) {
+	value, ok := c.m.Load(key)
+	if ok {
+		return value.(proxyLookupResult), true
+	} else {
+		return proxyLookupResult{}, false
+	}
+}
+
+// timedRefresh runs an endless loop, refreshing the given key
+// on a coarse-grained interval.  This is in case something changes in the user's network
+// configuration. E.g. they switch between wifi and wired if on a laptop.  Shouldn't be common
+// with AzCopy, but could happen so we may as well cater for it.
+func (c *proxyLookupCache) endlessTimedRefresh(key url.URL, representativeFullRequest *http.Request) {
+	if c.refreshInterval == 0 {
+		return
+	}
+
+	for {
+		time.Sleep(c.refreshInterval)
+		value := c.getProxyNoCache(representativeFullRequest)
+		c.m.Store(key, value)
+	}
+}
```

Comment:
This could be simplified with a time.Ticker and a list of URLs, potentially.

Reply:
Yes, I think if where were lots of URLs expected, the we'd have to do something like that.  Or maybe just "range" through the contents of the map.

In our case, where will usually only be 1 or 2 URLs in the map, so I'm leaning towards leaving it as-is.